buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java

```diff
@@ -440,9 +448,22 @@ protected synchronized PoolThreadCache initialValue() {
 
             Thread current = Thread.currentThread();
             if (useCacheForAllThreads || current instanceof FastThreadLocalThread) {
-                return new PoolThreadCache(
+                final PoolThreadCache cache = new PoolThreadCache(
                         heapArena, directArena, tinyCacheSize, smallCacheSize, normalCacheSize,
                         DEFAULT_MAX_CACHED_BUFFER_CAPACITY, DEFAULT_CACHE_TRIM_INTERVAL);
+
+                if (DEFAULT_CACHE_TRIM_RATE > 0) {
+                    EventExecutor executor = ThreadExecutorMap.currentExecutor();
+                    if (executor != null) {
```

Comment:
Is it possible that we have a null ?
in that case DEFAULT_CACHE_TRIM_RATE won't have effect, should we log a warning ?

Reply:
yes its possible if people call from there own thread which not power a `EventExecutor`. This is also why I think we should not log. If we log then it should be in trace I guess.

Reply:
In one of my use cases I am pretty sure that I am not inside an event executor.
For instance, I am inside the request handler of a Servlet on Tomcat, it is not a Netty managed thread, in this case when I call allocator#directBuffer(), this may trigger this method and I will end up in a PoolThreadCache which does not observe this limit.

So it is better not to log, you are right.

But is there a way to leverage this feature in this case? 

I guess I should wrap my allocator#directBuffer() call with a ThreadExecutorMap.apply(Runnable, EventExecutor)
But I still don't have an EventExecutor.

So my self answer is that I should manage a special global EventExecutor and wrap all of the calls

can you confirm @normanmaurer  ?

Reply:
@eolivelli in this case you will need to call `PooledByteBufAllocator.trimCurrentThreadCache()` from the thread from time to time to have the same effect. That said we already trim by ourself from time to time, so you only need to explicit do it by yourself if you not allocate frequently enough.

See https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PoolThreadCache.java#L192.

Alternative you can also disable the cache for threads that are not owned by netty:

https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L442

Reply:
@normanmaurer  thank you.
It is very clear to me now.